### PR TITLE
Add documentation for the Network resource

### LINKED ISSUE #1579: Documentation for `Network` resource
We are lacking documentation for the `extensions.gardener.cloud/v1alpha1.Network` resource in https://github.com/gardener/gardener/tree/master/docs/extensions. For all other extension contracts we already have extensive documentation, see https://github.com/gardener/gardener/tree/master/docs#extensions.

### PR DESCRIPTION
**Pull request by zanetworker**
**What this PR does / why we need it**:
Add documentation for the network resource. 

**Which issue(s) this PR fixes**:
Fixes #1579 

**Special notes for your reviewer**:

**Release note**:

```improvement developer
Documentation for the `Network` extension resource has been added. 
```

### CODE REVIEW

File: docs/extensions/network.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# Gardener Network Extension
+
+Gardener is an open-source project that provides a nested user model. Basically, there are two types of services provided by Gardener to its users:
+
+- Managed: users only request a Kubernetes cluster (Clusters-as-a-Service)
+- Hosted: users utilize Gardener to provide their own managed version of Kubernetes (Cluster-Provisioner-as-a-service)
+
+For the first set of users, the choice of network plugin might not be so important, however, for the second class of users (i.e., Hosted) it is important to be able to customize networking based on their needs.
+
+Furthermore, Gardener provisions clusters on different cloud-providers with different networking requirements. For example, Azure does not support Calico Networking [1], this leads to the introduction of manual exceptions in static add-on charts which is error prune and can lead to failures during upgrades.
```
Comment (vpnachev) [inline docs/extensions/network.md]:
```suggestion
Furthermore, Gardener provisions clusters on different cloud-providers with different networking requirements. For example, Azure does not support Calico Networking [1], this leads to the introduction of manual exceptions in static add-on charts which is error prone and can lead to failures during upgrades.
```

File: docs/extensions/network.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+Finally, every provider is different, and thus the network always needs to adapt to the infrastructure needs to provider better performance. Consistency does not necessarily lie in the implementation but in the interface.
```
Comment (vpnachev) [inline docs/extensions/network.md]:
```suggestion
Finally, every provider is different, and thus the network always needs to adapt to the infrastructure to provide better performance. Consistency does not necessarily lie in the implementation but in the interface.
```

Comment (vpnachev) [reply docs/extensions/network.md]: What is the purpose of the sentence  `Consistency does not necessarily lie in the implementation but in the interface.` ?

File: docs/extensions/network.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  podCIDR: 100.244.0.0/16
+  serviceCIDR: 100.32.0.0/13
+  type: calico
+  providerConfig:
+    apiVersion: calico.networking.extensions.gardener.cloud/v1alpha1
+    backend: bird
+    ipam:
+      cidr: usePodCIDR
+      type: host-local
+    kind: NetworkConfig
```
Comment (vpnachev) [inline docs/extensions/network.md]: I think that `kind` should be right after the `apiVersion` for better readability.

File: docs/extensions/network.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+```
+
+The above resources is divided into two parts:
+
+- global configuration (e.g., podCIDR, serviceCIDR, and type)
+- provider specific config (e.g., for calico we can choose to configure a `bird` backend)
```
Comment (vpnachev) [inline docs/extensions/network.md:50]: It will be nice if a link to this [docu](https://github.com/gardener/gardener-extensions/blob/master/controllers/networking-calico/docs/usage-as-end-user.md) is added here.

**Review by vpnachev — CHANGES_REQUESTED** [review]

**Comment by vpnachev** [comment]
Also refer this document from here https://github.com/gardener/gardener/blame/master/docs/README.md#L72

File: docs/extensions/network.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
+# Gardener Network Extension
+
+Gardener is an open-source project that provides a nested user model. Basically, there are two types of services provided by Gardener to its users:
+
+- Managed: users only request a Kubernetes cluster (Clusters-as-a-Service)
+- Hosted: users utilize Gardener to provide their own managed version of Kubernetes (Cluster-Provisioner-as-a-service)
```
Comment (rfranzke) [inline docs/extensions/network.md]: It's quite confusing what a "user" is. Can you clarify by using "users" for end-users (people only using shoot clusters but not operating/setting up Gardener itself) and "operators" for the other group?

File: docs/extensions/network.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+For the first set of users, the choice of network plugin might not be so important, however, for the second class of users (i.e., Hosted) it is important to be able to customize networking based on their needs.
```
Comment (rfranzke) [inline docs/extensions/network.md]: I would rephrase the first sentence - for many people it is important as it is running in their cluster and might provide higher levels of services (e.g., calico provides network policies while flannel does not).

For the second scenario ("hosted") it's quite unimportant - the only thing that the operators want to do is to offer choice for end-users. They themselves are not the primary users of shoots.

File: docs/extensions/network.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+Furthermore, Gardener provisions clusters on different cloud-providers with different networking requirements. For example, Azure does not support Calico Networking [1], this leads to the introduction of manual exceptions in static add-on charts which is error prune and can lead to failures during upgrades.
+
+Finally, every provider is different, and thus the network always needs to adapt to the infrastructure needs to provider better performance. Consistency does not necessarily lie in the implementation but in the interface.
+
+## Motivation
+
+Prior to the `Network Extensibility` concept, Gardener followed a mono network-plugin support model (i.e., Calico). Although this seemed to be the more stable approach, it did not completely reflect the real use-case.
```
Comment (rfranzke) [inline docs/extensions/network.md]:
```suggestion
Prior to the `Network Extensibility` concept, Gardener followed a mono network-plugin support model (i.e., Calico only). Although this seemed to be the more stable approach, it did not completely reflect the real use-case.
```

Comment (rfranzke) [reply docs/extensions/network.md]: I don't understand the second sentence. Why does this "seem to be the more stable approach"?

File: docs/extensions/network.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+The above resources is divided into two parts:
+
+- global configuration (e.g., podCIDR, serviceCIDR, and type)
+- provider specific config (e.g., for calico we can choose to configure a `bird` backend)
+
+> **Note**: certain cloud-provider extensions might have webhooks that would modify the network-resource to fit into their network specific context. As previously mentioned, Azure does not support IPIP, as a result, the [Azure provider extension](https://github.com/gardener/gardener-extensions/tree/master/controllers/provider-azure) implements a [webhook](https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-azure/pkg/webhook/network/mutate.go) to mutate the backend and set it to `None` instead of `bird`.
+
+## Supporting a new Network Extension Provider
+
+To add support for another networking provider (e.g., weave, Cilium, Flannel, etc.). A network extension controller needs to be implemented which would have its own custom configuration specified in the `spec.providerConfig` in the network resource. For example, if support for a network plugin named `gardenet` is required, the following `Network` resource would be created:
```
Comment (rfranzke) [inline docs/extensions/network.md]:
```suggestion
To add support for another networking provider (e.g., weave, Cilium, Flannel, etc.) a network extension controller needs to be implemented which would optionally have its own custom configuration specified in the `spec.providerConfig` in the `Network` resource. For example, if support for a network plugin named `gardenet` is required, the following `Network` resource would be created:
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: docs/extensions/network.md (hunk 2 above)
Comment (zanetworker) [reply docs/extensions/network.md]: Just to make it clear that we still maintain consistency (which Gardener promises), by provident a consistent interface.

File: docs/extensions/network.md (hunk 3 above)
Comment (zanetworker) [reply docs/extensions/network.md]: thats from a `kubectl get` output. Will fix.

File: docs/extensions/network.md (hunk 7 above)
Comment (zanetworker) [reply docs/extensions/network.md]: The intention here is to say, it was thought that it is easier to maintain and control.

File: docs/extensions/network.md (hunk 6 above)
Comment (zanetworker) [reply docs/extensions/network.md]: I am not sure you got my point in the text, so I rephrased it.

File: docs/extensions/network.md (hunk 7 above)
Comment (rfranzke) [reply docs/extensions/network.md]: Maybe you could re-formulate this. "more stable" sounds like what we are now doing is less stable.

**Review by rfranzke — APPROVED** [review]

**Review by vpnachev — APPROVED** [review]
/approve
